esmvaltool/preprocessor/_multimodel.py

```diff
@@ -13,42 +25,314 @@
 logger = logging.getLogger(__name__)
 
 
-def _put_in_cube(stats, stats_name, ncfiles, fname):
-    """quick cube building"""
-    stats_cube = iris.cube.Cube(stats, long_name=stats_name)
-    stats_cube.attributes['_filename'] = fname
-    stats_cube.attributes['NCfiles'] = str(ncfiles)
+def _time_pts_by_calendar(cube):
+    """Extract time points depending on calendar"""
+    # a_p is a prefactor that brings time points
+    # to 365-day calendar
+    time_units = cube.coord('time').units
+    if time_units.calendar is not None:
+        if time_units.calendar == '360_day':
+            a_p = 365. / 360.
+        else:
+            a_p = 1.
+    else:
+        logger.debug("No calendar type: assuming 365-day")
+        a_p = 1.
+    # return tpoints
+    t_points = [c * a_p for c in cube.coord('time').points]
+    return t_points
+
+
+def _plev_fix(dataset, pl_idx):
+    """Extract valid plev data
+
+    this function takes care of situations
+    in which certain plevs are completely
+    masked due to unavailable interpolation
+    boundaries.
+    """
+    if np.ma.is_masked(dataset):
+        # keep only the valid plevs
+        if not np.all(dataset.mask[pl_idx]):
+            statj = np.ma.array(dataset[pl_idx], mask=dataset.mask[pl_idx])
+        else:
+            logger.debug('All vals in plev are masked, ignoring.')
+            statj = None
+    else:
+        mask = np.zeros_like(dataset[pl_idx], bool)
+        statj = np.ma.array(dataset[pl_idx], mask=mask)
+
+    return statj
+
+
+def _compute_statistic(datas, name):
+    """Compute multimodel statistic"""
+    datas = np.ma.array(datas)
+    statistic = datas[0]
+
+    if name == 'median':
+        statistic_function = np.ma.median
+    elif name == 'mean':
+        statistic_function = np.ma.mean
+    else:
+        raise NotImplementedError
+
+    # no plevs
+    if len(datas[0].shape) < 3:
+        statistic = statistic_function(datas, axis=0)
+        return statistic
+
+    # plevs
+    for j in range(statistic.shape[0]):
+        plev_check = []
+        for cdata in datas:
+            fixed_data = _plev_fix(cdata, j)
+            if fixed_data is not None:
+                plev_check.append(fixed_data)
+        len_stat_j = sum(1 for _ in plev_check)
+        stat_all = np.ma.zeros((len_stat_j, statistic.shape[1],
+                                statistic.shape[2]))
+        for i, e_l in enumerate(plev_check):
+            stat_all[i] = e_l
+
+        # check for nr models
+        if len_stat_j >= 2:
+            statistic[j] = statistic_function(stat_all, axis=0)
```

Comment:
I think you can skip creating the `stat_all` array and directly use the `plev_check` list as input to `statistic_function`, like so:
```
if len(plev_check) > 1:
    statistic[j] = statistic_function(plev_check, axis=0)
```

Reply:
done! but with conversion of list plev_check to np.ma array -- dereferencing the list buys us memory; also np.ma.median is very fussy and needs it to force the masked array